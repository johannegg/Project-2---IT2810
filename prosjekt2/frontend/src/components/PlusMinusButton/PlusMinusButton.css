.playlist-modal-overlay {
	position: fixed;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	background: rgba(0, 0, 0, 0.5);
	display: flex;
	justify-content: center;
	align-items: center;
	z-index: 1000;
	cursor: default;
}

.playlist-modal-overlay.sidebar-open {
	top: -20px;
}

.playlist-modal-container {
	position: fixed;
	top: 50%;
	left: 50%;
	transform: translate(-50%, -50%);
	background: #ffffff;
	padding: 20px;
	width: 290px;
	max-width: 90%;
	border-radius: 12px;
	box-shadow: 0 4px 12px rgba(0, 0, 0, 0.15);
	text-align: center;
	overflow: hidden;
	word-wrap: break-word;
	overflow-y: auto;
	max-height: 80vh;
}

.playlist-modal-overlay.sidebar-open .playlist-modal-container {
	top: 15%;
}

.playlist-modal-container h3 {
	font-size: 1.2em;
	color: #3d3d3d;
	margin-bottom: 15px;
	padding: 0 10px;
	white-space: normal;
}

.playlist-selection {
	list-style: none;
	padding: 0;
	margin: 10px;
	max-height: 300px;
	overflow-y: auto;
}

.playlist-selection li {
	margin-bottom: 10px;
}

.playlist-selection button {
	width: 200px;
	padding: 10px;
	background-color: var(--theme-darker-green);
	border: none;
	color: black;
	border-radius: 5px;
	cursor: pointer;
	font-size: 1rem;
	transition:
		background-color 0.3s,
		transform 0.2s;
	overflow: hidden;
	text-overflow: ellipsis;
	white-space: nowrap;
}

.playlist-selection button:hover {
	background-color: var(--theme-even-darker-green);
	transform: scale(1.02);
	color: #ffffff;
}

.closeBtn {
	background-color: #ccc;
	color: #000;
	border: none;
	border-radius: 5px;
	cursor: pointer;
	padding: 8px 16px;
	font-size: 1rem;
	margin-top: 15px;
	transition: background-color 0.3s;
	display: inline-block;
}

.closeBtn:hover {
	background-color: #aaa;
}

.plusMinus-button {
	background: none;
	border: none;
	padding: 0;
	cursor: pointer;
	display: inline-flex;
	align-items: center;
	justify-content: center;
	outline: none;
	margin: 5px;
	font-size: 30px;
	position: relative;
}

.plusMinus-button::after {
	content: attr(data-label);
	position: absolute;
	top: 100%;
	left: 50%;
	transform: translateX(-50%);
	background-color: #fff;
	color: #555;
	padding: 4px 8px;
	border-radius: 4px;
	border: 1px solid #d4d4d4;
	white-space: nowrap;
	font-size: 12px;
	font-weight: bold;
	opacity: 0;
	pointer-events: none;
	transition: opacity 0.3s ease;
	z-index: 10;
}

.plusMinus-button:hover::after,
.plusMinus-button:focus::after {
	opacity: 1;
}

@media (prefers-color-scheme: dark) {
	.playlist-modal-container {
		background-color: #000;
	}

	.playlist-modal-container h3 {
		color: #ffffff;
	}

	.playlist-modal-overlay {
		background-color: rgba(28, 28, 28, 0.8);
	}

	.playlist-selection button {
		background-color: var(--theme-even-darker-green);
		color: #ffffff;
	}

	.playlist-selection button:hover {
		background-color: var(--theme-darker-green);
		color: #000;
	}

	.closeBtn {
		background-color: #aaa;
	}

	.closeBtn:hover {
		background-color: #ccc;
	}
}

@media (max-width: 650px) {
	.plusMinus-button {
		font-size: 25px;
	}
}
