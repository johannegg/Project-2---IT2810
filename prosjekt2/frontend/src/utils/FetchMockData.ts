export interface Song {
	id: number;
	title: string;
	artist: string;
	year: number;
	views: number;
	lyrics: string;
	genre: string;
}

export const fetchSongs = async (): Promise<Song[]> => {
	try {
		const response = await fetch("/mockdata.json");
		if (!response.ok) {
			throw new Error("Failed to fetch songs");
		}
		const data: Song[] = await response.json();
		return data;
	} catch (error) {
		console.error("Error fetching songs:", error);
		throw error;
	}
};
