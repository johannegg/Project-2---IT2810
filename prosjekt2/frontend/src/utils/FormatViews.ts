export const formatViews = (views: number): string => {
	return new Intl.NumberFormat("en", { notation: "compact", compactDisplay: "short" }).format(
		views,
	);
};
