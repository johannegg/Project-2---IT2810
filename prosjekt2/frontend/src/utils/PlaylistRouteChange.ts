import { PlaylistData } from "../pages/Playlists/Playlists";
import { NavigateFunction } from "react-router-dom";

export const routeChange = (playlist: PlaylistData, navigate: NavigateFunction) => {
  const path = `/playlist/${playlist.name.toLowerCase().replace(/ /g, "-")}`;
  navigate(path, { state: playlist });
};
