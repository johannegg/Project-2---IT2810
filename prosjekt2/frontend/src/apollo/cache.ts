import { InMemoryCache, makeVar } from "@apollo/client";
import { SongData } from "../utils/types/SongTypes";
import { PlaylistData } from "../pages/Playlists/Playlists";

// Initialize sessionStorage and localStorage if values are not already set
const initializeStorage = (key: string, defaultValue: string) => {
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, defaultValue);
  }
};

const initializeLocalStorage = (key: string, defaultValue: string) => {
  if (!localStorage.getItem(key)) {
    localStorage.setItem(key, defaultValue);
  }
};

// Initialize values in storage
initializeStorage("minViews", "0");
initializeStorage("maxViews", "1000000");
initializeStorage("selectedGenres", "[]");
initializeStorage("sortOption", "views_desc");
initializeStorage("homeSearchTerm", "");
initializeStorage("favoritesSearchTerm", "");
initializeLocalStorage("favoriteSongs", "[]");
initializeLocalStorage("playlists", "[]");

// Load saved values from storage
const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
const savedMinViews = Number(sessionStorage.getItem("minViews"));
const savedMaxViews = Number(sessionStorage.getItem("maxViews"));
const savedSortOption = sessionStorage.getItem("sortOption") || "views_desc";
const savedHomeSearchTerm = sessionStorage.getItem("homeSearchTerm") || "";
const savedFavoritesSearchTerm = sessionStorage.getItem("favoritesSearchTerm") || "";
const savedFavoriteSongs = JSON.parse(localStorage.getItem("favoriteSongs") || "[]");
const savedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");

// Create reactive variables
export const genreFilterVar = makeVar<string[]>(savedGenres);
export const sortOptionVar = makeVar<string>(savedSortOption);
export const minViewsVar = makeVar<number>(savedMinViews);
export const maxViewsVar = makeVar<number>(savedMaxViews);
export const homeSearchTermVar = makeVar<string>(savedHomeSearchTerm);
export const favoritesSearchTermVar = makeVar<string>(savedFavoritesSearchTerm);
export const favoriteSongsVar = makeVar<SongData[]>(savedFavoriteSongs);
export const playlistsVar = makeVar<PlaylistData[]>(savedPlaylists);  
export const songDataVar = makeVar<SongData[]>([]);

// Configure cache with reactive variables
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        genreFilter: {
          read() {
            return genreFilterVar();
          },
        },
        sortOption: {
          read() {
            return sortOptionVar();
          },
        },
        minViews: {
          read() {
            return minViewsVar();
          },
        },
        maxViews: {
          read() {
            return maxViewsVar();
          },
        },
        homeSearchTerm: {
          read() {
            return homeSearchTermVar();
          },
        },
        favoritesSearchTerm: {
          read() {
            return favoritesSearchTermVar();
          },
        },
        favoriteSongs: {
          read() {
            return favoriteSongsVar();
          },
        },
        playlists: {
          read() {
            return playlistsVar();
          },
        },
        songData: {
          read() {
            return songDataVar();
          },
        },
      },
    },
  },
});

// Sync reactive variables with sessionStorage and localStorage
homeSearchTermVar.onNextChange((newHomeSearchTerm) => {
  sessionStorage.setItem("homeSearchTerm", newHomeSearchTerm);
});

favoritesSearchTermVar.onNextChange((newFavoritesSearchTerm) => {
  sessionStorage.setItem("favoritesSearchTerm", newFavoritesSearchTerm);
});

genreFilterVar.onNextChange((newGenres) => {
  sessionStorage.setItem("selectedGenres", JSON.stringify(newGenres));
});

minViewsVar.onNextChange((newMinViews) => {
  sessionStorage.setItem("minViews", String(newMinViews));
});

maxViewsVar.onNextChange((newMaxViews) => {
  sessionStorage.setItem("maxViews", String(newMaxViews));
});

sortOptionVar.onNextChange((newSortOption) => {
  sessionStorage.setItem("sortOption", newSortOption);
});

favoriteSongsVar.onNextChange((newFavorites) => {
  localStorage.setItem("favoriteSongs", JSON.stringify(newFavorites));
});

playlistsVar.onNextChange((newPlaylists) => {
  localStorage.setItem("playlists", JSON.stringify(newPlaylists));
  window.dispatchEvent(new Event("storage")); 
});

export default cache;
