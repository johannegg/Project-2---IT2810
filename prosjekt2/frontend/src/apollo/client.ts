import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
