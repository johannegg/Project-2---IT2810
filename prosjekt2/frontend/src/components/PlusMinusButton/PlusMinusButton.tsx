import React, { useState, useCallback } from "react";
import { FiPlusCircle, FiMinusCircle } from "react-icons/fi";
import { SongData } from "../../utils/types/SongTypes";
import { PlaylistData } from "../../pages/Playlists/Playlists";
import { useReactiveVar} from "@apollo/client";
import { playlistsVar, isSidebarOpenVar } from "../../apollo/cache";
import "./PlusMinusButton.css";

type PlusMinusButtonProps = {
	song: SongData;
	isInPlaylist?: boolean;
	playlistId?: string;
	onSongRemoved?: () => void;
};

const PlusMinusButton: React.FC<PlusMinusButtonProps> = ({
	song,
	isInPlaylist,
	playlistId,
	onSongRemoved,
}) => {
	const playlists = useReactiveVar(playlistsVar);
	const isSidebarOpen = useReactiveVar(isSidebarOpenVar);
	const [showModal, setShowModal] = useState(false);
	const [feedbackMessage, setFeedbackMessage] = useState("");

	const debounce = (func: () => void, delay: number) => {
		let timer: NodeJS.Timeout;
		return () => {
			clearTimeout(timer);
			timer = setTimeout(() => {
				func();
			}, delay);
		};
	};

	const clearFeedbackMessage = useCallback(
		debounce(() => setFeedbackMessage(""), 3000),
		[],
	);

	const handleAddSongToPlaylist = (playlistId: string) => {
		let songAdded = false;

		const updatedPlaylists = playlists.map((playlist: PlaylistData) => {
			if (playlist.id === playlistId) {
				const isSongAlreadyInPlaylist = playlist.songs.some((s) => s.id === song.id);
				if (!isSongAlreadyInPlaylist) {
					songAdded = true;
					return { ...playlist, songs: [...playlist.songs, song] };
				} else {
					setFeedbackMessage("Song is already in playlist.");
					clearFeedbackMessage();
					return playlist;
				}
			}
			return playlist;
		});

		if (songAdded) {
			setFeedbackMessage("Song successfully added!");
			clearFeedbackMessage();
			playlistsVar(updatedPlaylists);
		}
	};

	const handleRemoveSongFromPlaylist = () => {
		if (playlistId) {
			const updatedPlaylists = playlists.map((playlist: PlaylistData) => {
				if (playlist.id === playlistId) {
					return { ...playlist, songs: playlist.songs.filter((s) => s.id !== song.id) };
				}
				return playlist;
			});
			playlistsVar(updatedPlaylists);
			if (onSongRemoved) onSongRemoved();
		}
	};

	const toggleModal = () => setShowModal(!showModal);

	return (
		<>
			<button
				className="plusMinus-button"
				data-label={isInPlaylist ? "Remove song" : "Add song"}
				onClick={(e) => {
					e.stopPropagation();
					if (isInPlaylist) {
						handleRemoveSongFromPlaylist();
					} else {
						toggleModal();
					}
				}}
			>
				{isInPlaylist ? <FiMinusCircle color="#afc188" /> : <FiPlusCircle color="#afc188" />}
			</button>

			{showModal && (
				<div className={`playlist-modal-overlay ${isSidebarOpen ? 'sidebar-open' : ''}`}  onClick={(e) => e.stopPropagation()}>
					<div className="playlist-modal-container">
						<h3>Select a playlist to add "{song.title}"</h3>
						{feedbackMessage && (
							<label
								className={`feedback-message ${feedbackMessage === "Song successfully added!" ? "success" : "error"}`}
							>
								{feedbackMessage}
							</label>
						)}
						<ul className="playlist-selection">
							{playlists.map((playlist: PlaylistData) => (
								<li key={playlist.id}>
									<button
										onClick={(e) => {
											e.stopPropagation();
											handleAddSongToPlaylist(playlist.id);
										}}
									>
										{playlist.name + " " + playlist.icon}
									</button>
								</li>
							))}
						</ul>
						<button
							onClick={(e) => {
								e.stopPropagation();
								toggleModal();
							}}
							className="closeBtn"
						>
							Close
						</button>
					</div>
				</div>
			)}
		</>
	);
};

export default PlusMinusButton;
