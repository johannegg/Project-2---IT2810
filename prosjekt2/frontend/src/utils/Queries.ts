import { gql } from "@apollo/client";

export const GET_SONGS = gql`
	query GetSongs(
		$skip: Int
		$limit: Int
		$genres: [String]
		$sortBy: SortBy
		$searchTerm: String
		$minViews: Int
		$maxViews: Int
	) {
		songs(
			skip: $skip
			limit: $limit
			genres: $genres
			sortBy: $sortBy
			searchTerm: $searchTerm
			minViews: $minViews
			maxViews: $maxViews
		) {
			id
			title
			views
			year
			lyrics
			artist {
				name
				id
			}
			genre {
				name
			}
		}
	}
`;

export const GET_SONG_COUNT = gql`
  query GetSongCount(
    $genres: [String]
    $searchTerm: String
    $minViews: Int
    $maxViews: Int
  ) {
    songCount(
      genres: $genres
      searchTerm: $searchTerm
      minViews: $minViews
      maxViews: $maxViews
    )
  }
`;

export const GET_GENRE_COUNTS = gql`
  query GetGenreCounts(
    $searchTerm: String
    $minViews: Int
    $maxViews: Int
    $genres: [String]
  ) {
    genreCounts(
      searchTerm: $searchTerm
      minViews: $minViews
      maxViews: $maxViews
      genres: $genres
    ) {
      name
      count
    }
  }
`;

export const CREATE_USER = gql`
	mutation CreateUser($username: String!) {
		createUser(username: $username) {
			id
			username
			favoriteSongs {
				id
				title
				views
				year
				lyrics
				artist {
					name
				}
				genre {
					name
				}
			}
			playlists {
				id
				name
				songs {
					id
					title
					views
					year
					lyrics
					artist {
						name
						id
					}
					genre {
						name
					}
				}
			}
		}
	}
`;

export const DELETE_USER = gql`
	mutation DeleteUser($username: String!) {
		deleteUser(username: $username)
	}
`;

export const ADD_FAVORITE_SONG = gql`
	mutation AddFavoriteSong($username: String!, $songId: ID!) {
		addFavoriteSong(username: $username, songId: $songId)
	}
`;

export const REMOVE_FAVORITE_SONG = gql`
	mutation RemoveFavoriteSong($username: String!, $songId: ID!) {
		removeFavoriteSong(username: $username, songId: $songId)
	}
`;

export const CREATE_PLAYLIST = gql`
	mutation CreatePlaylist(
		$username: String!
		$name: String!
		$backgroundcolor: String!
		$icon: String!
	) {
		createPlaylist(
			username: $username
			name: $name
			backgroundcolor: $backgroundcolor
			icon: $icon
		) {
			id
			name
			backgroundcolor
			icon
			songs {
				id
				title
			}
		}
	}
`;

export const DELETE_PLAYLIST = gql`
	mutation DeletePlaylist($username: String!, $playlistId: ID!) {
		deletePlaylist(username: $username, playlistId: $playlistId)
	}
`;

export const ADD_SONG_TO_PLAYLIST = gql`
	mutation AddSongToPlaylist($username: String!, $playlistId: ID!, $songId: ID!) {
		addSongToPlaylist(username: $username, playlistId: $playlistId, songId: $songId) {
			id
			name
			backgroundcolor
			icon
			songs {
				id
				title
			}
		}
	}
`;

export const REMOVE_SONG_FROM_PLAYLIST = gql`
	mutation RemoveSongFromPlaylist($username: String!, $playlistId: ID!, $songId: ID!) {
		RemoveSongFromPlaylist(username: $username, playlistId: $playlistId, songId: $songId) {
			id
			name
			backgroundcolor
			icon
			songs {
				id
				title
			}
		}
	}
`;
