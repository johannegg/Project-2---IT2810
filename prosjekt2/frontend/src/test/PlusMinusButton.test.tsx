import { render, screen, fireEvent } from "@testing-library/react";
import PlusMinusButton from "../components/PlusMinusButton/PlusMinusButton";
import { playlistsVar } from "../apollo/cache";
import { ADD_SONG_TO_PLAYLIST, REMOVE_SONG_FROM_PLAYLIST } from "../utils/Queries";
import { MockedProvider } from "@apollo/client/testing";
import "@testing-library/jest-dom";
import { describe, test, vi, beforeEach, beforeAll, expect } from "vitest";
import { waitFor } from "@testing-library/react";

// Mocking Apollo cache
vi.mock("../apollo/cache", async () => {
	const { makeVar } = await import("@apollo/client");
	return {
		playlistsVar: makeVar([]),
		isSidebarOpenVar: makeVar(false),
	};
});

describe("PlusMinusButton Component", () => {
	// Mock data for a song and playlist
	const mockSong = {
		id: "1",
		title: "Test Song",
		artist: { id: "1", name: "Test Artist" },
		views: 1000,
		year: 2023,
		genre: { name: "Test Genre" },
		lyrics: "La la la...",
	};
	const mockPlaylist = {
		id: "playlist1",
		name: "Test Playlist",
		songs: [mockSong],
		backgroundcolor: "#ffffff",
		icon: "ðŸŽµ",
	};

	// Mock GraphQL responses
	const mocks = [
		{
			// Mocking the ADD_SONG_TO_PLAYLIST mutation
			request: {
				query: ADD_SONG_TO_PLAYLIST,
				variables: { username: "testUser", playlistId: "playlist1", songId: "1" },
			},
			result: { data: { addSongToPlaylist: { success: true } } },
		},
		{
			// Mocking the REMOVE_SONG_FROM_PLAYLIST mutation
			request: {
				query: REMOVE_SONG_FROM_PLAYLIST,
				variables: { username: "testUser", playlistId: "playlist1", songId: "1" },
			},
			result: { data: { removeSongFromPlaylist: { success: true } } },
		},
	];

	beforeAll(() => {
		// Mocking the window.alert function
		window.alert = vi.fn();
	});

	beforeEach(() => {
		// Resetting mocks and reactive variables before each test
		vi.resetAllMocks();
		playlistsVar([mockPlaylist]); // Setting initial state for playlists
		localStorage.setItem("profileName", "testUser");
	});

	test("renders the add button by default", () => {
		// Render component with MockedProvider for GraphQL
		const { asFragment } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<PlusMinusButton song={mockSong} />
			</MockedProvider>,
		);

		// Assert that the "Add song" button is rendered
		expect(screen.getByRole("button", { name: "Add song" })).toBeInTheDocument();

		// Take a snapshot of the rendered component
		expect(asFragment()).toMatchSnapshot();
	});

	test("renders the remove button if isInPlaylist is true", () => {
		// Render component with isInPlaylist set to true
		const { asFragment } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<PlusMinusButton song={mockSong} isInPlaylist playlistId="playlist1" />
			</MockedProvider>,
		);

		// Assert that the "Remove song" button is rendered
		expect(screen.getByRole("button", { name: "Remove song" })).toBeInTheDocument();

		// Take a snapshot of the rendered component
		expect(asFragment()).toMatchSnapshot();
	});

	test("displays modal when add button is clicked", () => {
		const { asFragment } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<PlusMinusButton song={mockSong} />
			</MockedProvider>,
		);

		// Simulate a click on the "Add song" button
		fireEvent.click(screen.getByRole("button", { name: "Add song" }));

		// Assert that the modal is displayed
		expect(screen.getByText('Select a playlist to add "Test Song"')).toBeInTheDocument();
		expect(screen.getByRole("button", { name: "Close" })).toBeInTheDocument();

		// Take a snapshot of the modal
		expect(asFragment()).toMatchSnapshot();
	});

	test("adds song to a playlist when a playlist button is clicked", async () => {
		const { asFragment } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<PlusMinusButton song={mockSong} />
			</MockedProvider>,
		);

		// Simulate a click on the "Add song" button
		fireEvent.click(screen.getByRole("button", { name: "Add song" }));

		// Simulate a click on a playlist button in the modal
		fireEvent.click(screen.getByRole("button", { name: "Test Playlist ðŸŽµ" }));

		// Assert that the success feedback is displayed
		expect(await screen.findByText("Song successfully added!")).toBeInTheDocument();

		// Take a snapshot of the feedback message
		expect(asFragment()).toMatchSnapshot();
	});

	test("removes song from playlist when remove button is clicked", async () => {
		const { asFragment } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<PlusMinusButton song={mockSong} isInPlaylist playlistId="playlist1" />
			</MockedProvider>,
		);

		// Simulate a click on the "Remove song" button
		fireEvent.click(screen.getByRole("button", { name: "Remove song" }));

		// Wait for the reactive variable to be updated
		await waitFor(() => {
			const updatedPlaylists = playlistsVar();
			const playlist = updatedPlaylists.find((p) => p.id === "playlist1");
			expect(playlist?.songs).not.toContainEqual(mockSong);
		});

		// Take a snapshot of the feedback message
		expect(asFragment()).toMatchSnapshot();
	});

	test("alerts user if not logged in when adding a song", () => {
		// Simulate a logged-out user
		localStorage.removeItem("profileName");

		const { asFragment } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<PlusMinusButton song={mockSong} />
			</MockedProvider>,
		);

		// Simulate a click on the "Add song" button
		fireEvent.click(screen.getByRole("button", { name: "Add song" }));

		// Assert that an alert is triggered
		expect(window.alert).toHaveBeenCalledWith("You need to be logged in to add songs to playlists");

		// Take a snapshot of the component in the logged-out state
		expect(asFragment()).toMatchSnapshot();
	});
});
