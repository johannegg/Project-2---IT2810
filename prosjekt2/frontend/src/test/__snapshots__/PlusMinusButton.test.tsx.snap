// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlusMinusButton Component > adds song to a playlist when a playlist button is clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Add song"
    class="plusMinus-button"
  >
    <svg
      color="#afc188"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(175, 193, 136);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="16"
      />
      <line
        x1="8"
        x2="16"
        y1="12"
        y2="12"
      />
    </svg>
  </button>
  <div
    class="playlist-modal-overlay "
  >
    <div
      class="playlist-modal-container"
    >
      <h3>
        Select a playlist to add "Test Song"
      </h3>
      <label
        class="feedback-message error"
      >
        Song is already in playlist.
      </label>
      <ul
        class="playlist-selection"
      >
        <li>
          <button>
            Test Playlist ðŸŽµ
          </button>
        </li>
      </ul>
      <button
        class="closeBtn"
      >
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlusMinusButton Component > alerts user if not logged in when adding a song 1`] = `
<DocumentFragment>
  <button
    aria-label="Add song"
    class="plusMinus-button"
  >
    <svg
      color="#afc188"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(175, 193, 136);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="16"
      />
      <line
        x1="8"
        x2="16"
        y1="12"
        y2="12"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`PlusMinusButton Component > displays modal when add button is clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Add song"
    class="plusMinus-button"
  >
    <svg
      color="#afc188"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(175, 193, 136);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="16"
      />
      <line
        x1="8"
        x2="16"
        y1="12"
        y2="12"
      />
    </svg>
  </button>
  <div
    class="playlist-modal-overlay "
  >
    <div
      class="playlist-modal-container"
    >
      <h3>
        Select a playlist to add "Test Song"
      </h3>
      <ul
        class="playlist-selection"
      >
        <li>
          <button>
            Test Playlist ðŸŽµ
          </button>
        </li>
      </ul>
      <button
        class="closeBtn"
      >
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlusMinusButton Component > removes song from playlist when remove button is clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Remove song"
    class="plusMinus-button"
  >
    <svg
      color="#afc188"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(175, 193, 136);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="8"
        x2="16"
        y1="12"
        y2="12"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`PlusMinusButton Component > renders the add button by default 1`] = `
<DocumentFragment>
  <button
    aria-label="Add song"
    class="plusMinus-button"
  >
    <svg
      color="#afc188"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(175, 193, 136);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="16"
      />
      <line
        x1="8"
        x2="16"
        y1="12"
        y2="12"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`PlusMinusButton Component > renders the remove button if isInPlaylist is true 1`] = `
<DocumentFragment>
  <button
    aria-label="Remove song"
    class="plusMinus-button"
  >
    <svg
      color="#afc188"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(175, 193, 136);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="8"
        x2="16"
        y1="12"
        y2="12"
      />
    </svg>
  </button>
</DocumentFragment>
`;
